beaker-vue/src/lib/UINotebook.vue

```diff
@@ -0,0 +1,342 @@
+
+const isDragEnabled = computed(() => session.notebook?.cells.length > 1);
+const cellCount = computed(() => session.notebook?.cells?.length || 0);
+
+
+/**
+ * Modifies array in place to move a cell to a new location (or any other elem)
+ **/
+function arrayMove(arr, old_index, new_index) {
+    arr.splice(new_index, 0, arr.splice(old_index, 1)[0]);
+}
+
+// TODO move selectCellByObj, addCell, removeCell to common fns file
+//      this is repeated in NotebookControls
+//      Alternatively, expose these actions from lib/BeakerSession
+function selectCellByObj(cellObj) {
+  const index = session.notebook.cells.indexOf(cellObj);
+  props.selectCell(index);
+}
+
+const addCell = (toIndex) => {
+    const newCell = session.addCodeCell("");
+
+    if (typeof toIndex === 'number') {
+        // Move cell to indicated index
+        arrayMove(session.notebook.cells, cellCount.value - 1, toIndex)
+    }
+
+    selectCellByObj(newCell);
+
+    // Ensure cell is focused, not the editor or contents within
+    nextTick(() => {
+        focusSelectedCell();
+    });
+}
+
+const removeCell = () => {
+    session.notebook.removeCell(props.selectedCellIndex);
+
+    // Always keep at least one cell. If we remove the last cell, replace it with a new empty codecell.
+    if (cellCount.value === 0) {
+        session.addCodeCell("");
+    }
+
+    // Fixup the selection if we remove the last item.
+    if (props.selectedCellIndex >= cellCount.value) {
+      props.selectCell(cellCount.value - 1);
+    }
+};
+
+const commonSelectAction = (event) => {
+    const { target } = event;
+
+    const isEditingCode = target.className === 'cm-content'; // codemirror
+    const isTextArea = target.className.includes('resizeable-textarea');
+
+    if (isEditingCode || isTextArea) {
+        return false;
+    }
+
+    return true;
+};
+
+
+const selectNextCell = (event) => {
+    if (event && !commonSelectAction(event)) {
+        return;
+    }
+
+    const currentIndex = props.selectedCellIndex;
+    if (currentIndex === cellCount.value - 1) {
+        return;
+    }
+    props.selectCell(currentIndex + 1);
+
+    nextTick(() => {
+        focusSelectedCell();
+    });
+
+    if (event) {
+        event.preventDefault();
+    }
+};
+
+const selectPreviousCell = (event) => {
+
+    if (event && !commonSelectAction(event)) {
+        return;
+    }
+
+    const currentIndex = props.selectedCellIndex;
+    if (currentIndex === 0) {
+        return;
+    }
+    props.selectCell(currentIndex - 1);
+
+    nextTick(() => {
+        focusSelectedCell();
+    });
+
+    if (event) {
+        event.preventDefault();
+    }
+};
+
+function handleMoveCell(fromIndex, toIndex) {
+    arrayMove(session.notebook.cells, fromIndex, toIndex)
+    props.selectCell(toIndex);
+}
+
+/**
+ * Sets call to item being moved
+ * As well as dataTransfer so that drop target knows which one was dropped
+ **/
+function handleDragStart(event, beakerCell, index)  {
```

Comment:
Had to remove the types, temporarily, while the build fails for whatever reason? Typescript not working on `.lib/` folder. Can add back once we figure that out.